C++/source/number_theory/fibonacci_efficient.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+    Fibonacci number
+    -----------------
+    Find the N-th Fibonacci number, given the value of N.
+    (The Fibonacci sequence is characterized by the fact that every number
+    after the first two is the sum of the two preceding ones.)
+
+    Time complexity
+    ---------------
+    O(log N), where N is the term of the Fibonacci sequence to calculate
+
+    Space complexity
+    ----------------
+    O(1).
+*/
+
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+typedef unsigned long long int ULL;
+
+const int MAX_N = 93;   // fibonacci(94) goes beyond the range of ULL
+
+// Product of two matrices. Their sizes must be [n, m] and [m, p]
+vector< vector<ULL> > matrix_product(const vector< vector<ULL> > &first_matrix, const vector< vector<ULL> > &second_matrix) {
+    // Example: If First[n, m] and Second[m, p], then their product will be Product[n, p]
+    vector< vector<ULL> > product_of_matrices(first_matrix.size(), vector<ULL> (second_matrix[0].size()));
+
+    for (int i = 0; i < (int) first_matrix.size(); i++) // For previous example i:1..n
+            for (int j = 0; j < (int) second_matrix[0].size(); j++) // For previous example j:1..p
+                for (int k = 0; k < (int) second_matrix.size(); k++) // For previous example k:1..m
```

Comment:
`i`, `j` and `k` should be of type `size_t`.

Also, the comments say `i: 1..n`, `j: 1..p` etc. which should be `i: 0..n-1` etc.